# Regional DNA methylation analysis using DMRcate  
Using data preprocessed in our script:  
 meth01_process_data.R & meth02_process_data.R 

```{r setdir03, echo = F}
knitr::opts_knit$set(root.dir = "../")
```

 we have already set up our analysis

```{r }
if(!exists("pheno")){
  source("code/meth02_analyze_data.R")
}
```

Load package for regional analysis "DMRcate"
 see [Peters et al. Bioinformatics 2015](https://epigeneticsandchromatin.biomedcentral.com/articles/10.1186/1756-8935-8-6).  
Other popular options for conducting Regional DNA methylation analysis in R are Aclust and bumphunter 

```{r }
suppressMessages(library(DMRcate)) # Popular package for regional DNA methylation analysis
```

First we need to define a model

```{r }
model <- model.matrix(~as.factor(pheno$Smoke)+
                      as.factor(pheno$Sex)+
                      as.numeric(pheno$Age)+
                      as.numeric(pheno$CD8T)+
                      as.numeric(pheno$NK)+
                      as.numeric(pheno$Bcell)+
                      as.numeric(pheno$Mono)+
                      as.numeric(pheno$Gran))
```

Regions are now agglomerated from groups of significant probes 
Let's run the regional analysis using the Beta-values from our preprocessed data

```{r }
myannotation <- cpg.annotate("array", betas.clean, analysis.type="differential",arraytype="EPIC",
                             what="Beta",design=model, coef=2)
```

We don't find any significant regions (FDR<0.05), So let's try a simpler model as an example

```{r }
model <- model.matrix(~as.factor(pheno$Smoke)+
                      as.factor(pheno$Sex)+
                      as.numeric(pheno$Age))

myannotation <- cpg.annotate("array", betas.clean, analysis.type="differential",arraytype="EPIC",
                             what="Beta",design=model, coef=2)
```

Regions are now agglomerated from groups of significant probes 
where the distance to the next consecutive probe is less than lambda nucleotides away

```{r }
dmrcoutput.smoking <- dmrcate(myannotation, lambda=1000, C=2)
```

Let's look at the results

```{r }
head(dmrcoutput.smoking$results)
```

Visualizing the data can help us understand where the region lies 
relative to promoters, CpGs islands or enhancers
Let's extract the genomic ranges and annotate to the genome

```{r }
results.ranges <- extractRanges(dmrcoutput.smoking, genome = "hg19")
```

Plot the DMR using the Gviz
if you are interested in plotting genomic data the Gviz is extremely useful
Let's look at the first region

```{r }
results.ranges[1]
pheno$Smoke <- ifelse(pheno$Smoke==1, "Smoker", "Non-Smoker")
groups <- c(Smoker="magenta", Male="Non-Smoker")
cols <- groups[as.character(pheno$Smoke)]
```{r fig.width=9, fig.height=6, dpi=300}
DMR.plot(ranges=results.ranges, dmr=1, CpGs=betas.rcp, what="Beta", phen.col=cols, genome="hg19", arraytype = "EPIC")
```

Extracting CpGs-names and locations

```{r }
chr <- gsub(":.*", "", dmrcoutput.smoking$results$coord[1])
start <- gsub("-.*", "", gsub(".*:", "", dmrcoutput.smoking$results$coord[1]))
end <- gsub(".*-", "", dmrcoutput.smoking$results$coord[1])
```

CpG ID and individual metrics

```{r }
cpgs <- dmrcoutput.smoking$input[dmrcoutput.smoking$input$CHR %in% chr & dmrcoutput.smoking$input$pos >= start & dmrcoutput.smoking$input$pos <=end,]
knitr::kable(cpgs[1:4,])
```

End of script 03


---
title: "meth03_regional_analysis.R"
author: "acardena"
date: "Thu Jun 15 17:34:45 2017"
---
